junit-jupiter-engine/src/main/java/org/junit/jupiter/engine/discovery/AbstractOrderingVisitor.java

```diff
@@ -166,8 +149,33 @@ private boolean canOrderWrappers() {
 			return this.orderingAction != null;
 		}
 
-		private void orderWrappers(List<WRAPPER> wrappers) {
-			this.orderingAction.accept(wrappers);
+		private void orderWrappersWithOrderFrom(Set<WRAPPER> wrappers) {
+			List<WRAPPER> orderedWrappers = new ArrayList<>(wrappers);
+			this.orderingAction.accept(orderedWrappers);
+			Set<Object> distinctOrderedWrappers = new LinkedHashSet<>(orderedWrappers);
+
+			int difference = orderedWrappers.size() - wrappers.size();
+			int distinctDifference = distinctOrderedWrappers.size() - wrappers.size();
+			if (difference > 0) { // difference >= distinctDifference
+				logDescriptorsAddedWarning(difference);
+			}
+			if (distinctDifference < 0) { // distinctDifference <= difference
+				logDescriptorsRemovedWarning(distinctDifference);
+			}
+
+			orderWrappersWithOrderFrom(wrappers, distinctOrderedWrappers);
+		}
+
+		@SuppressWarnings("unchecked")
+		private void orderWrappersWithOrderFrom(Set<WRAPPER> wrappers, Set<Object> orderedWrappers) {
+			// Avoid ClassCastException if a misbehaving ordering action added a non-WRAPPER
+			for (Object wrapper : orderedWrappers) {
+				//noinspection SuspiciousMethodCalls
+				if (wrappers.remove(wrapper)) {
+					// guaranteed by wrappers.contains
+					wrappers.add((WRAPPER) wrapper);
```

Comment:
Not happy with this yet. A misbehaving ordering action could add test descriptors with a broken equals/hashcode implementation.

Reply:
This comment is outdated, right?

Reply:
Yes!